### ISSUE TITLE
Official dark theme has much worse performance than every other theme

### ISSUE THREAD

**Issue by Tokinsom** [Scirra/Construct-3-bugs#3017]
## Problem description

After playing around with every available theme (official & 3rd party) I've discovered that the official C3 dark theme performs significantly slower than the others. This is most apparent in the event sheet editor,  and with UI effects / animations enabled, but can be noticed elsewhere too. This has been noted as a browser tech issue in the forums (something with chrome layouts?) but wouldn't that suggest all themes would be affected?

## Attach a .c3p

Blank .c3p with ~300 events

## Steps to reproduce

1. Just about any event sheet function

## Observed result

There is noticeable delay opening / closing / transitioning between ACE windows & editors,..pasting or moving event blocks, conditions, and actions...opening/closing context menus, etc.

## Expected result

No delays at all, like the other themes.

## More details

**Affected browsers/platforms:**

All

**First affected release:**

All

## System details

<details><summary>View details</summary>

Platform information
Browser: NW.js
Browser version: 0.35.1
Browser engine: Chromium
Browser architecture: 64-bit
Context: nwjs
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 4
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (@ac9418ba9c3bd7f6baaffa0b055dfe147e0f8364) NWjs/0.35.1 (like Chrome/71.0.3578.80)
C3 release: r152 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 267 gb
Storage usage (approx): 1.4 gb (0.5%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (NVIDIA GeForce GTX 1080 Ti Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by AshleyScirra** [comment]
Sounds like the same issue as #2177, and I'm not sure there's much we can do about it:

> I'm not sure if there's much we can do about this. If Edge is OK, then presumably it's some Chrome-specific issue with something about the dark theme CSS, and we don't really have any good tools to figure out what part of the CSS is problematic. So it might mean you just have to choose another theme or browser.

The previous issue mentioned Edge was OK and the issue seemed to be Chrome specific. Did you test every browser? Are there really no differences between them?

As ever we also cannot reproduce what you are seeing unless you provide the project you tested with.

**Comment by alastajj** [comment]
I've been noticing (148 stable) that it's been sluggish lately too, but I wondered oh it might just be my computer or because my game is big? 

In my case, if I restart Construct it starts performing faster again, and sometimes also fixes a glitch with dragging events around (it starts scrolling up automatically when dragging a condition). But I don't if it's just me or not.

**Comment by SnipG** [comment]
I tested long time ago this dialog open time in Dark Theme.  When you have huge events with 300+ actions, then dialogs open 3x longer then default theme/or my own theme that i made(which modifies 1500+css line). I  tried to delete some CSS from dark theme that affect Dialogs, which changed nothing visually and dialog open got like 40% faster, but  but it was still very slow. So it could be possible to speed it up. But it would need a alot of time to do "trial and error testing".

**Comment by AshleyScirra** [comment]
Please note that to make progress on issues, it is helpful to answer any questions we ask - so far nobody has answered mine.

**Comment by Tokinsom** [comment]
@AshleyScirra I just tested in Chrome, Firefox, and Edge browsers.

Chrome: Same performance in browser VS desktop
Firefox: Even worse performance VS desktop, and crashed when expanding all events on a single sheet (578)
Edge: Consistently crashed when opening the project due to "long script". 3 attempts were made.

I can't upload my project publicly - is there an email I can send it to for testing? All you need is a project with a lot of events, anyway...

Also I have to say this does not speak well at all to using browser tech for the editor. My game is hardly into the "medium scale" category and has a long ways to go yet.

**Comment by alastajj** [comment]
@Tokinsom the program only becomes unusable for me if I expand all the events in my game, but I only have 1800 events in my primary sheet so far though. However when I keep most of the sheet collapsed, it works pretty smoothly. The normal theme does seem a little faster than the dark one for me.

**Comment by mOOnpunk** [comment]
Same experience as alastajj. To be honest i dont like the dark theme, but there is simply no other option if you want to save your eyes. I wish there was one like the default theme, but a darker grey instead of black.

**Comment by Tokinsom** [comment]
@mOOnpunk I suggested a dark theme without any stylistic alterations over a year ago. Sadly didn't get much traction https://construct3.ideas.aha.io/ideas/C3-I-470

**Comment by shortercode** [comment]
I'm working on a revision of my "rainbow tabs" theme that is dark. The intention is to be fairly minimal. It's not an official theme, and also it's quite a lot of work creating a dark theme so it's taking awhile.

**Comment by alastajj** [comment]
@Tokinsom  yeah those are all excellently stated issues, a clear reflection of what many people have been feeling about dark theme. The progress on getting these basic issues fixed is taking many years!

It seems like a good idea to just use the same style as the default theme, it's surprising that they didn't do it that way to begin with!

**Comment by mOOnpunk** [comment]
@Tokinsom i just voted for it. Hope it happens.

**Comment by AshleyScirra** [comment]
> Chrome: Same performance in browser VS desktop

What do you mean by "VS desktop"? Are you testing the desktop build of Construct? That wouldn't be any different, and you said the same thing for Firefox where it is not applicable.

**Comment by Tokinsom** [comment]
@AshleyScirra I tested in all the browsers listed & desktop version of C3.

Chrome, Firefox, and Desktop version seemed to perform similarly, whereas Edge was slow and stopped responding when opening large projects.

There have never been performance issues with the default theme in any of these browsers.

The takeaway is, yes, the dark theme performs worse in all scenarios VS other themes.

There are even similar 3rd party dark themes that run fine compared to the default one. There _is_ something in the official dark theme slowing it down.

**Comment by AshleyScirra** [comment]
I still don't understand parts of what you've described, for example:

> Firefox: Same performance VS desktop

There is no Firefox version of the desktop app. What did you compare it to? It doesn't make sense to me. I need to properly understand what the situation is before we can proceed.

**Comment by Tokinsom** [comment]
@AshleyScirra Desktop = C3 Desktop Version.

I compared browser performance with the C3 desktop performance. I wasn't sure if there would be a difference.

I also compared performance between the browsers with Dark & Default themes.

Again, there is a noticeable performance hit with the dark theme regardless of C3 version or Browser, when compared to the default theme or any other theme, including 3rd party.

I seriously feel like you're just giving me the round-about here.

**Comment by AshleyScirra** [comment]
OK, thanks for clarifying. It's just the comment seemed confusing and it's important to have clear feedback to be able to make sense of bug reports.

**Comment by Tokinsom** [comment]
Is there anything more I can help with? The "need info" tag has not been removed and no mention of this in the last few updates.

**Comment by AshleyScirra** [comment]
Could you attach a project demonstrating the issue so I can make sure I am seeing what you are seeing? The performance could easily depend on the specific types of events used. (Our guidelines require this, because it's always useful.)

Adding the 'difficult' tag because I can't see any good way of investigating this without a lengthy trial-and-error process, and there is over 5000 lines of CSS for the dark theme.

**Comment by LukeWolfGames** [comment]
@AshleyScirra Since this issue is impacting me when I am making my game, I have been looking into it. I hope I can be of assistance. Please have a look at this c3p file:

https://www.dropbox.com/s/n1y7urma5jvaku0/Slow%20Object%20Search%20in%20Larger%20Event%20Sheets.c3p?dl=0

### Here's my analysis of the problem across Construct 3 Desktop, Chrome and Firefox:

**The event block example I used to produce the problem and this block is included in the event sheets to test performance of the editor across different platforms and themes:**
**Condition:** Sprite is overlapping sprite 
**Action:** Sprite spawns another sprite

Event sheet 1 has 1000 events in Group1, 50 events in Group2 and 0 events in Group3.
Event sheet 2 has 100 events.
Event sheet 3 has 50 events.
Event sheet 4 has 0 events.

**The platforms I tested with:**
Construct 3 Desktop
Chrome
FireFox

**Themes I tested:**
I tested all themes (Default, Light and Dark) across these platforms

**General observations across all themes and platforms:**
1. Event sheet 1 takes a long time to expand it's Group1. Once open, it does seem to slow down the editor when the events are visible on the screen even if they are not in the view port range and the user has the scroll to see events further down. As long as they are visible in the editor, it seems to impact on Construct's performance, making it slow and jerky when moving events around, or opening a dialogue box to edit or add a condition or action.
2. Performance is not as badly impacted if the events are not visible but hidden in a group that is not expanded but there is still some noticeable performance hindered here.
3. Check the other events sheets, the editor. You will notice the editor is not slowed down as much when editing those event sheets in the above manner I specified, especially the empty event sheet.
4. Different themes and platforms seem to handle this differently. The platform that handles this the worst is Construct 3 Desktop using the Dark Theme. The platform that handles this the best is Firefox with any theme. However, Firefox is still impacted by this problem.

**Now I will go into more detail with what I was doing to make these observations:**
1. On Event sheet 1: All groups are closed. Open one of the groups and observe how long it takes to open. It is recommended to keep the other 2 groups closed when doing this so you can better understand the lag difference.
2. Once you have a group open, observe how it takes to do the following actions:
2a. Moving events by clicking and dragging their position
2b. Double clicking an event or action to open up the dialogue box.
2c. Typing in the dialogue box's search bar to search for objects.
2d. picking and navigating the object picker window and various other actions during the process of selecting an object, picking a condition or action, setting it and closing the window.
2e. Editing items in the project bar.
3. Test these actions on the other event sheets and take note of the difference in performance.

### Specific observations I've made while doing the above testing for each platform and theme:

**CONSTRUCT 3 DESKTOP (The one I mainly use):**

**DEFAULT THEME:**

**Opening event groups:**
1. Opening Group1 on Event sheet 1 is slow in comparison to opening the other groups.
2. Group2 opens reasonably fast with a short but tolerable pause.
3. Group3 opens the moment it is clicked with no pause.

**Moving events (clicking and dragging):**
1. Moving events in Group1 has a noticeable pause showing the reconstruction of the event block in it's new location before it is visible.
2. Moving events in Group2 is a lot quicker than Group1

**Opening Dialogue Box:**
1. Opening a dialogue box by pressing E is faster than double clicking a condition or action and opening it that way.
2. There is no pause when opening the dialogue box with only Group2 or GRoup3 of Event Sheet 1 open.

**Closing dialogue box:**
1. There is a noticeable pause when closing a dialogue box while Group1 of Event sheet 1 is viewed.
2. There is little to no pause when closing a dialogue box with only Group2 or Group3 when viewing Event sheet 1.

**Typing:**
1. Typing input is lagged when Event Sheet 1 is viewed. This extends to other parts of the editor, for example: if you try to type to rename an event sheet or an object in the project bar, the typing is lagged.
2. The typing is not lagged when viewing Event sheet 2, Event sheet 3 or Event sheet 4.

**Navigating the dialogue box:**
1. There are noticeable pauses whenever picking an object, condition, action or loading other areas of the dialogue box where you are asked to input values (like setting angle for example), just general moving around the different areas of the dialogue box seems to have this effect when Group1 of Event Sheet 1 is open.
2. Doing point 1 but with only Group 2 or Group3 open has little to no lag at all.
3. There is no lag when performing point 1 on the other event sheets.

**Scrolling:**
1. Scrolling is lagged when only Group1 of Event Sheet 1 is open.
2. Scrolling is not lagged when only Group2 of Event Sheet 1 is open.
3. Scrolling is not lagged when scrolling on the other event sheets.

**DARK THEME:**

**Opening event groups:**
1. Group1 of Event sheet 1 viewed only: Seems a little slower than the default theme.
2. Group2 of Event sheet 1 viewed only: Pause seems the same as default theme.
3. Group3 of Event sheet 1 viewed only: instantly opens like default theme. 

**Moving events (clicking and dragging):**
1. Group1 of Event sheet 1 viewed only: Much slower than default theme. What's new here is that you can notice a light blue line that appears before a event reconstructs itself in the new location.
2. Group2 of Event sheet 1 viewed only: A bit slower than default theme. What's new here is that you can notice a light blue line that appears before a event reconstructs itself in the new location. This is faster than Group1, though.

**Opening Dialogue Box:**
1. Group1 of Event sheet 1 viewed only: Much slower than default theme.
2. Group2 of Event sheet 1 viewed only: A bit slower than the default theme.
3. Group3 of Event sheet 1 viewed only: Opens instantly.

**Closing dialogue box:**
Not as slow as opening dialogue box but slower than the default theme. Mainly noticeable in Group1, a bit slower than the default theme in Group2 and Group3 closes instantly.

**Typing:**
1. Typing is lagged when Event Sheet 1 is viwed, event sheet 2. It is slower when only group1 is open, a little faster with only group 2 is open and little improvement but still lagged when group3 is the only one open. 
2. Typing Input is definitely lagged on other areas of the editor when Event sheet 1 is viewed.

**Navigating the dialogue box:**
1. Group1 of Event sheet 1 viewed only: Navigation when selecting objects, picking conditions actions, or any other navigation with the dialogue box option is much slower than the default theme and impacts user experience.
2. Group2 of Event sheet 1 viewed only: A very short delay during navigation.
3. Group3 of Event sheet 1 viewed only: Instant response.
4. Event Sheet 2 has a short pause.
5. Event sheet 3 has a short pause.
6. Event Sheet 4 has an instant response.

**Scrolling:**
1. Group1 of Event sheet 1 viewed only: As slow as default theme.
2. Group2 of Event sheet 1 viewed only: Instant scrolling response.
3. Other Event Sheets have instant scrolling response.

**LIGHT THEME:**

**Opening event groups:**
Same as dark theme across all areas.

**Moving events (clicking and dragging):**
1. Group1 of Event sheet 1 viewed only: Slower than dark theme.
2. Group2 of Event sheet 1 viewed only: Same as default and dark theme.

**Opening Dialogue Box:**
1. Group1 of Event sheet 1 viewed only: Slower than dark theme.
2. Group2 of Event sheet 1 viewed only: Same as dark theme.
3. Group3 of Event sheet 1 viewed only: Instantly opens.

**Closing dialogue box:**
1. Group1 of Event sheet 1 viewed only: Slower than dark theme.
2. Group2 of Event sheet 1 viewed only: Same as dark theme.
3. Group3 of Event sheet 1 viewed only: Instantly closes.

**Typing:**
Same as dark theme across all areas.

**Navigating the dialogue box:**
Seems slower than dark theme for Group1 of Event Sheet1.
Response is the same for the otehr areas of the editor, apart from that.

**Scrolling:**
Same as the dark theme across all areas.

**BROWSERS:**
Main differences compared to the desktop version when testing Group 1 in particular and Group 2 of event Sheet 1. It is safe to say at this point that the other groups and event sheets are not really a problem due to their size.

**CHROME:**

**DEFAULT THEME:**
1. Typing is still lagged but not as badly. It seems to be slower with Group1 open than when it is closed though.
2. Moving events is slower than the desktop version.
3. Opening and closing dialogue boxes is still slow but not as jerky.
4. Scrolling is a little faster than in the Desktop version.

** DARK THEME:**
1. Typing was very slow when typing in a dialogue box to begin with, but then it smooths out to a more normal rate but still lagged.
2. Moving events is slower than the default theme.
3. Opening and closing dialogue box takes a long time compared to the default theme.
4. Navigating the dialogue box is much slower than the default theme.
5. Scrolling is similar to the default theme, a little faster than the desktop version.

**LIGHT THEME:**
1. Typing still lagged but not as badly as the desktop version, faster than the dark theme but about the same as the default theme.
2. Moving events is slower than the dark theme.
3. Opening and closing dialogue box takes a long time compared to the dark theme.
4. Navigating the dialogue box is much slower than the dark theme.
5. Scrolling is similar to the default theme, a little faster than the desktop version.

**FIREFOX:**

**DEFAULT THEME:**
1. Typing is instant, no lag at all.
2. Opening Group1 of Event sheet 1 is just as laggy as the other browsers.
3. Moving events is a little faster than Chrome but still lagged.
4. Opening and closing dialogue boxes is faster than chrome but still lagged.
5. Navigating the dialogue box is much faster than Chrome with little delay.
6. Scrolling instant with no lag at all.

**DARK THEME:**
1. Typing is instant, no lag at all.
2. Opening Group1 of Event sheet 1 is just as laggy as the other browsers.
3. Moving events is a bit slower than the default theme.
4. Opening and closing dialogue boxes is the same as the default theme.
5. Navigating the dialogue box is the same as the default theme.
6. Scrolling instant with no lag at all.

**LIGHT THEME:**
1. Typing is instant, no lag at all.
2. Opening Group1 of Event sheet 1 is much slower than Default and Dark theme.
3. Moving events is a bit slower than the default theme.
4. Opening and closing dialogue boxes is the same as the default theme.
5. Navigating the dialogue box is the same as the default theme.
6. Scrolling instant with no lag at all.

Please let me know if you need more information.

**Comment by luckyrawatlucky** [comment]
@LukeWolfGames there is another dark theme available in CONSTRUCT 3 ADDONS page [Theme Dark](https://www.construct.net/en/make-games/addons/297/theme-dark)

**Comment by LukeWolfGames** [comment]
@luckyrawatlucky Thanks for sharing, I'll check it out but I believe Scirra will only want to test the performance of Construct with the official themes that are built in and not any addons provided by third parties.

**Comment by LukeWolfGames** [comment]
Just a question in reference to #3292 - why can't you guys mimic how Construct 2 adjusts event sheet columns? Because the event sheet performance was really good in Construct 2.

**Comment by AshleyScirra** [comment]
It's a completely different architecture. C3 uses HTML and CSS. C2 uses tons of hand-written C++ layout code which was hard work to write and difficult to maintain, but faster at layout. There is almost nothing transferrable between the two, they are polar opposites.

**Comment by Tokinsom** [comment]
Any updates on this? I know it's difficult and you mentioned the dark theme has 5,000 lines but to be fair this only pertains to the event sheet editor so that should narrow it down quite a lot already.
@LukeWolfGames has provided the most detailed .c3p and breakdown I think I've ever seen on this bug report platform.

This is an official theme and a very big factor in C3's overall UX.

I started writing a keymapping/gamepad config in a new .c3p and I'm already getting ~0.5 second delays in the event editor at a measly 146 events!

**Comment by LukeWolfGames** [comment]
Yes, it would be good to get an update on this issue as this really impacts the user experience. It does seem like splitting your events into different event sheets can help larger projects with this problem for the interim. I look forward to Scirra improving Construct 3's performance.

@Tokinsom Thanks for the recognition! I've recently discovered that unless the customer base provides accurate and helpful information to resolve problems, Scirra does not fix them. Also, since it is sometimes difficult to know exactly what is considered accurate and helpful information as a customer, I have deducted that the only way to get problems resolved is to spend the time to provide detailed analysis. Unfortunately, this level of support from the customer base is unreliable and I still recommend that Scirra hire professionally trained testers to do this job.

**Comment by AshleyScirra** [comment]
This is one of the most difficult and time-consuming issues ever filed to the tracker. The 'difficult' tag has been added to reflect this. I appreciate your patience and recognition that this is likely a long-term project to investigate.

I must emphasise that if you confuse this issue with other ones it becomes even more difficult to deal with. There are two issues that have been mentioned here: firstly relating to poor performance when using the default theme (or any theme), and secondly relating to poor performance in the dark theme specifically relative to the default theme (i.e. dark theme has a performance impact). These are different problems. Reading through long and complex comments such as posted by @LukeWolfGames earlier which mix up these two problems, and trying to unpick one problem from it, adds even more difficulty and slow-down to deal with an already difficult and time-consuming issue. It is of very significant importance to make sure you only discuss one problem at a time; it really makes a difference to how easy it is to investigate and deal with reports, especially when they are so difficult as this one. By this point it would be a lot easier to deal with this if we started over with a new issue with concise information about a single problem all covered in the original report. Keeping all relevant information concise and in one place is a big help when returning to an issue periodically with often long gaps in between, during which it's easy to forget what the status of the issue is, or where to collect the various relevant bits of information from.

The performance issue ultimately comes down to layout performance in the browser, i.e. working out where to position elements. This is largely determined by the browser engine and the amount of content currently shown, but is affected by CSS. The reason this only applies to expanded events is because collapsed events are hidden and don't need time spent processing their layout, hence it only applying to the amount of content shown.

Some of the difficult factors that are going to make it hard to make progress on this include:

- This ultimately comes down to layout performance of the browser engine, which is out of our hands.
- In some cases, the browser does a full-document layout. Sometimes it shouldn't be doing this, so is a kind of browser bug. However when this does happen, all CSS in the editor is involved - not just the event sheet view.
- There are sophisticated tools to delve in to the performance of JavaScript. However none exist for layout.
- There is lots of documentation on how to avoid performance pitfalls in JavaScript. Virtually none exists for layout.
- The dark theme was made by our designer. Making changes involves familiarizing ourselves with someone else's work (and a large amount of work at that), and carefully reviewing any changes to ensure they don't regress the appearance of the dark theme.

In short there is little we can do other than make shotgun (random) changes and hope it makes for a measurable improvement. This is potentially incredibly time consuming and even so is unclear it will arrive at a solution. As a tiny team dealing with thousands of bugs, hundreds of feature requests, ongoing user support, and everything else involved in running a business, with an issue that has no obvious solution, no documentation to help us, no tools to help analyse it, an unbounded amount of time that could be spent on it, and no guarantee of a solution, it is clearly going to be difficult to make progress in a reasonable amount of time.

I think the most likely solution will come from browser engine optimisations that speed up the existing layout & CSS sufficiently that it is no longer a serious problem. I tried Chrome Canary and it already appears a lot better. Firefox has a different browser engine too and recently introduced CSS containment which is useful for optimising layout performance (and C3 is already jam-packed with loads of CSS containment in an effort to optimise layout), so is worth comparing that.

For the record the desktop download of C3 is essentially identical to Chrome, so there's no point testing it. The most useful things to test are:

- Chrome stable browser
- Chome canary browser (I've already seen a big performance improvement here)
- Firefox stable browser
- Firefox nightly browser

The canary/nightly results will indicate any browser performance optimisations that help and will arrive on the stable channel in the next few months. Chrome Canary in particular is worth keeping an eye on since Google are currently in the process of rewriting the entire layout engine in the LayoutNG project, which has the potential to bring major performance improvements (and this may be why Canary already seems to be a lot faster).

If you still want us to do further investigation on this ourselves, it would be helpful to:
1) As noted, be clear which particular problem you are talking about (general performance vs. dark theme specifically)
2) Focus on a specific single user action, to narrow the scope of what we are investigating to just one thing
3) Try to be concise. Basically, short focused reports are easier to deal with.

Then we can potentially try to figure out how to optimise this on a case-by-case basis. I still cannot guarantee any results though. I've noticed one possibly related issue is that showing the resize cursors in the event sheet view can be slow, which appears to come down to a [Chrome browser bug](https://bugs.chromium.org/p/chromium/issues/detail?id=664066); given that specific case has a clear identification and I can think of some alternative approaches, it may be possible to work around. So I'll focus on that particular case for the time being.

**Comment by AshleyScirra** [comment]
I hacked together a workaround for the cursor performance issue, but early indications are it's no better in the current version of Chrome and significantly worse on the current version of Firefox, but much faster in both Chrome Canary and Firefox Nightly. So I think this particular aspect of the issue will have to wait until the faster versions of Chrome and Firefox come out (Chrome 79 and Firefox 71). Marking as blocked until then.

**Comment by LukeWolfGames** [comment]
Before we start getting pedantic about things, let’s read into things correctly. 
First, I recommend reviewing the title of this bug report, it says: 

“Official dark theme has much worse performance than every other theme #3017”

This means this is one issue.
Second, I was answering your previous question on the 20th of August which was:

“Could you attach a project demonstrating the issue so I can make sure I am seeing what you are seeing?”

Your comment, here, along with many others in this thread, communicated that you are very confused about what the issue actually is. Now, @Tokinsom did answer this before on the 9th of July:

“Chrome: Same performance VS C3 desktop
Firefox: Same performance VS C3 desktop / Chrome Browser, and stopped responding when expanding all events on a single sheet (578)
Edge: Consistently stopped responding when opening the project due to "long script". 3 attempts were made. Performance was noticeably bad just going through the settings with default theme...but Edge is crap so this is expected.”

But since you still couldn’t see what the problem was, I answered that question in more detail and provided the c3p file.

So, since you are comparing “performance between the dark theme and every other theme”, my ‘long and complex comments’ provided feedback and detailed analysis for this issue. In order to prove there is much worse performance on the dark theme, I had to investigate this by testing each of the themes and provide feedback so you can see what the problem is. Just like you requested. I am not the one adding layers of complication to an already complex problem when things like that and this are happening mid-discussion:

![](https://i.imgur.com/upiVE9E.png)

It’s very important that you not only pay attention to what your customers are saying but read into things correctly so you can arrive at the correct solution. It’s fine to not understand something and request more information, but it’s not fine when someone answers and you decide to lecture them about adding complexity after you showed a considerable escalating confusion.  That just gets people's backs up like so. You also mention later on that you had no documentation or tools to analyse it. I provided this on the 25th of August. It should give you something that you can work off, and if it’s not something that you can resolve on the existing platform, perhaps you should consider moving to another platform. There is no easy and straightforward answer to any of this since it is an involved issue, and as much as we hope that you find a solution to the problem, we can’t make it any easier than making clear what the problem is we are experiencing. You are just going to have to think about it and this is something you have to understand.

As far as being a tiny team facing a huge workload, running a business and having no time yourself; we are your customers. We pay premium price to use your product. Additionally for us, here, we take out our precious limited time to give you feedback on what’s going wrong with it because there are aspects we do like about it and we want to see it improve. We don’t care about your internal problems so it would be very good if you stopped lumping them on us. I’ve already told you the solution about handling these issues; hire more people to help with the load - trained professional testers that know how to give you the correct information, more developers to help with fixing them, more customer support people, etc. Your customer base is not a reliable means of achieving these accurately every time and as your business grows, the volume of handling these issues will become worse and the problems will spread because you can’t physically handle it. Surely, with all the money you are getting from your customer base you can reinvest it back into the business to get additional support and make room to grow and improve because you're not going to find it by expecting your customers to be testers, developers and customer support people.

**Comment by mOOnpunk** [comment]
>As far as being a tiny team facing a huge workload, running a business and having no time yourself; we are your customers. We pay premium price to use your product. Additionally for us, here, we take out our precious limited time to give you feedback on what’s going wrong with it because there are aspects we do like about it and we want to see it improve. We don’t care about your internal problems so it would be very good if you stopped lumping them on us. 

This. It's the get out excuse for any situation at Scirra. Its also the reason i no longer file bug reports or post suggestions.

**Comment by AshleyScirra** [comment]
I was asked to give an update on the issue. I spent some time trying to set out some of the deep complexities involved and how customers could co-operate to best assist in arriving at a solution to the problem. I apologise if I appeared to blame any customer for anything - it was not my intention. When dealing with bug reports about half the job is training people how to provide useful information. I try to point out the ways people can present more useful information that increases the chance of the issue being resolved. That was the approach I was trying to take.

I should clarify one issue: when talking about tools and documentation for investigating the issue, I am referring to the browser developer tools and developer documentation. These are the things we routinely use to deal with user-reported issues. It is still the case that no such materials exist to my knowledge and that none have been highlighted in this thread.

We have dealt with thousands of issues on this issue tracker. The vast majority of them are dealt with swiftly, demonstrating our commitment to do everything it takes to provide the best quality product possible to our paying customers. It should not be difficult to accept the fact that some issues are more difficult than others, and in rare cases, very much so. It's unrealistic to portray us as having unlimited resources to expend, or that any particular solution is quick or easy or obvious. Some problems are deeply complicated and we have limited resources to expend on them. Making stronger demands does nothing to solve any of these issues and in fact will make it harder to arrive at a solution. I am asking for a little trust that I am presenting a reasonable portrayal of the situation from our side, and highlighting what can be done to increase the chance of a resolution.

I fear that this issue will fall down a path that has happened once or twice in the past, where people make increasingly strong demands that are simply unrealistic. There is no good way for us to handle that so the end result was to close the issue and stop investigating it. If you wish to avoid that outcome I'd ask for your co-operation and respect for the [Forum & Community guidelines](https://www.construct.net/en/forum/general/open-topic-33/forum-community-guidelines-141035).

Towards avoiding that outcome and also as I was previously suggesting, I believe the best way to make progress on this issue will be to close this issue and open specific, targeted issues that deal with particular user actions on a case-by-case basis, including all relevant information in the original post so it's in one place, including testing of Chrome stable and canary, and Firefox stable and nightly, with precise time measurements where possible so we have a way to measure any improvements. Bear in mind this will still be a long-term project to deal with and I cannot guarantee what kind of results it will produce. I am not currently in the office so I will look in to taking steps towards that next week.

**Comment by Tokinsom** [comment]
@Ashley I imagine if we make a new bug report it will literally be a copy/paste of @LukeWolfGames 's post on Aug. 25th.

There is only one bug report and one issue here. We are all talking about the performance of the dark theme. That much has been clear from the start.

What happened was @LukeWolfGames went above and beyond to provide you with a large amount of information regarding the performance of the dark theme -- and while he was at it -- also provided some comparative performance measures with the default and light themes.

It did not then diverge into multiple bug reports / issues.
This has only spiraled out of control because more and more information and specifics were posted at your behest, which is now seemingly moot based on your latest responses, which is incredibly frustrating because we've been at it for 6 months now and no one at Scirra has even lifted a finger to actually look into it!!

You mentioned that your designer (Paulo?) created the dark theme. Would it not be wise to consult with him? Perhaps he knew something he was doing with the dark theme was a bit taxing, but did it anyway? **Perhaps since he created it, he should be the one investigating and fixing/maintaining it?**

You also mentioned chrome layouts could be the real issue and that this is just a symptom. Wouldn't that mean the default theme would have major performance issues like the dark theme as well? Because it does not. This heavily implies it's an issue with the dark theme itself. Even if Canary provides slightly better performance with the dark theme (it does) it seems akin to buying a more powerful machine instead of writing more optimized code (i.e. the default theme will still be faster than dark theme).

**Comment by LukeWolfGames** [comment]
This is the summary of this issue in its relevance so far:

@Tokinsom reported official dark theme has much worse performance than every other theme.
@AshleyScirra requested a comparison and a .c3p file.
@alastajj and @SnipG confirmed the issue based on reproducible data.
@Tokinsom provided the comparison feedback that @AshleyScirra requested.
@alastajj and @mOOnpunk provide further confirmation as to the cause of the issue and more comparison feedback.
@AshelyScirra requested clarification as to what “desktop version” meant.
@Tokinsom clarified what he meant by “desktop version”.
@AshleyScirra requested clarification as to what “desktop version” meant.
@Tokinsom clarified what he meant by “desktop version”.
@Tokinsom requested what additional information was needed since no mention of this in updates and no response for 31 days.
@AshelyScirra requested a project demonstrating the issue so he could see what we were seeing and tagged this as a difficult issue.
@LukeWolfGames attached a project demonstrating the issue along with detailed analysis of the performance amongst all themes to investigate the issue with the dark theme.
@Tokinsom requested an update since no mention of this in updates and no response to issue for 1 month and 25 days.
@AshleyScirra split the one issue into two separate issues then indicated that the analysis @LukeWolfGames provided had two functions: mixing up the two issues and adding more difficulty to the issue, additionally went into detail of the complexity of the issue, introduced internal business issues such as staffing levels, volume of requests including feature requests, user support and all unforeseen things involved in running a business, justifying lack of time, and requested us to explain what the problem was in a concise format.
@LukeWolfGames clarified the reason for providing the .c3p file and analysis linking it to on-topic discussion, clarified that what was requested was already delivered, and disputed customer accountability in terms of internal business issues and recommended solutions for those issues.
@AshleyScirra clarified his approach on the issue, made an external reference to bug report statistics, predicted that that customers were making increasingly strong and unrealistic demands, requested customer co-operation and referenced the forum rules and community guidelines and indicated to close the bug report issue and start all over again.
@Tokinsom clarified that if customers made a new bug report, it will be copy/paste of the detailed analysis that @LukeWolfGame’s provided, clarified that there is only one issue and that creating a new bug report will mean less information since customers have already submitted their feedback regarding this issue and linked a possible resolution suggestion.

Customers have already done their job in providing what you have requested. 

Requesting performance improvements when there is a clear issue with performance impacting user experience is not unreasonable or unrealistic.
Workload, staffing levels and internal business problems that you keep referencing are irrelevant to the issue and it is not something that customers can help with.
You also need to understand that the repercussions of refusing to investigate issues reported about your product (especially performance) are going to have a much more significant impact on your business than it is your customers in the grand scheme of things.

Closing this issue to make more is going to make things even more complicated. You were just going on about having information in one place now you want information in all different places. I think this issue should stay open, if not for the fact that I don’t really want to post my analysis in multiple threads but also for simplification reasons. The thing is anything relating to how you want to handle personal organisation of information you collect from your customers is up to you. Sure, we’re happy to follow your guidelines as much as possible but expecting your customers to post the same issue over and over again is unreasonable. It just disrespects our time.

**Comment by SnipG** [comment]
This issue should just be seperated to:

- Issue report to get Dark theme faster, almost same as default theme. 

- Feature request to improve default theme and looks if there i way to improve how themes work. Users made themes(also includes dark theme) should not override official css in external file, but should built into main one. Adds way to remove and change just 1 css file. And on top of it optimize it. If someone has bigger project, then he should not suffer because some css made for mobile etc drags it down.
C3 pushes browsers limits so should we have way to optimize stuff.
 
Some points:
Canarly does not seems to bring help!
Performance does seems to be changing, depending on example.
For example: With plank 1k event test in dark theme, firefox had no issues while chrome did. While i remember with other examples it was similar bad performance as chrome. So it most likley invloves multiple issues internally which can impact css layout perf, but i am not sure 100% because of its chaotic nature and  because it must involve massive amount of comparisons runs, which iam not up to do.
Not to mention just some simple case that easily affects testing, when you open event and expression dictonary also opens in dark theme, it adds like +50%-100 opening time instantly. So if you test sometimes with and without it, results easily change. Not to mention some other options affecting it.

**Comment by AshleyScirra** [comment]
A key misunderstanding here is, to be honest, from our side and what we need when we look in to reports, the type of information here is not very helpful. The earlier long comment mixes up maybe a dozen different bugs, all intermingled in different ways. Also, a key point is going in to reams of detail usually makes it harder to fix a problem, not easier. As an extreme example, imagine dumping 500 pages of documents on someone's desk and saying "fix it" - it's far, far more work to sort that out than look in to a concise and accurate paragraph summing up a specific problem. It is of key importance that information is concise and focused to one narrow topic for normal issues; for issues such as complex and difficult as this one, it's the difference between whether anything can be done at all or not. I've been trying to point out what needs to change to improve the chance that we can actually do something about this.

There are still unrealistic proposals being thrown about here. For example, Paulo's role is non-technical and this is a deeply complex technical problem, so it would not be appropriate to ask him to fix it.

I can see how this would be frustrating if you think this report clearly describes a specific issue, or that there is anything obvious to be done about it. Unfortunately neither is the case.

As I promised, I'm back in the office and making an attempt to split this out in to separate issues to improve the chance we can make some kind of progress on this. However I must emphasise that I cannot guarantee anything since this largely stems from the deep internals of browser engines, which we have limited control over. My findings so far are this:

- There is a specific problem that the way we change the mouse cursor is unreasonably slow in Chrome, due to a bug. It happens in both themes. I've filed #3426 to track that. As yet another complicating factor, this can often crop up while trying to test other cases, potentially interfering with the results, so care must be taken to avoid the mouse cursor changing while investigating other cases.
- Opening dialogs, such as the 'Add action' dialog, does appear to be slower in the dark theme than the light theme. I've filed #3427 to track that.
- Scrolling a long event sheet view (with care taken to avoid the mouse cursor changing) appears to be somewhat slower in the dark theme. I've filed #3428 to track that. However the difference may be explained by the fact the dark theme has more advanced styling so requires more paint work.

I could not reproduce some other cases described. Using Chrome Canary 80.0.3946.0 with the dark theme:
- dragging and dropping events seems reasonably fast, providing the cursor does not change
- typing in the search field in the 'Add action' dialog seems reasonably fast
- adding a subfolder in the project bar seems reasonably fast

If these can be reproduced under certain circumstances, these are likely separate cases that need to be tracked separately as well.

There is little more that can be done with this particular issue report. I've done my best to split off individual cases to investigate in the previously mentioned issues. Further, since tempers seem to be fraying (and this also will continue to make it even harder to deal with this report), I do think it has served its purpose, so I am closing it and locking it from further commenting. If you think other cases not already covered by the issues I've filed so far need investigating, please file them separately, using the most concise and focused approach you can - I've tried to write the issues I've filed in the style that is most helpful to demonstrate this. And once again I must emphasise this is an incredibly difficult type of problem to deal with; there are still no easy solutions and I still cannot guarantee when or if any progress will be made. As I pointed out earlier, the most likely solution is that browser engine optimisations speed up things as they are, although with a great deal of effort and hard work, we may be able to chip away at the performance problem in a few specific cases.
